test/gtest/cpu_multi_head_attention.hpp

```diff
@@ -0,0 +1,414 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include "tensor_holder.hpp"
+#include "conv_tensor_gen.hpp"
+#include <type_traits>
+
+#include <miopen/miopen.h>
+#include <gtest/gtest.h>
+#include <miopen/layernorm.hpp>
+
+#include "tensor_holder.hpp"
+#include "cpu_layernorm.hpp"
+#include "get_handle.hpp"
+#include "../driver/tensor_driver.hpp"
+#include "verify.hpp"
+#include <random>
+
+namespace test {
+namespace cpu {
+
+template <typename T>
+auto squareRoot(T value) -> decltype(std::sqrt(value))
+{
+    static_assert(std::is_arithmetic<T>::value, "squareRoot only supports arithmetic types");
+    return std::sqrt(value);
+}
+
+template <typename T>
+void BrodcastParMul(tensor<T> A_mat, tensor<T> B_mat, tensor<T>& ret_value)
+{
+    // std::mutex coutMutex;
+    size_t A_col = A_mat.desc.GetLengths()[3];
+    ret_value.par_for_each([&](size_t b_id, size_t h_id, size_t s_id, size_t d_id) {
+        T sum(0);
+
+        for(size_t pd_id = 0; pd_id < A_col; ++pd_id)
+        {
+            // std::lock_guard<std::mutex> guard(coutMutex);
+            sum += A_mat(b_id, 0, s_id, pd_id) * B_mat(0, h_id, pd_id, d_id);
+        }
+        ret_value(b_id, h_id, s_id, d_id) = sum;
+    });
+}
+
+template <typename T>
+void ParMul(tensor<T> A_mat, tensor<T> B_mat, tensor<T>& ret_value, size_t key_dim = 1)
+{
+    std::mutex coutMutex;
+    size_t A_col = A_mat.desc.GetLengths()[3];
+    ret_value.par_for_each([&](size_t b_id, size_t h_id, size_t s_id, size_t d_id) {
+        T sum(0);
+        std::lock_guard<std::mutex> guard(coutMutex);
```

Comment:
If `ret_value.par_for_each` access to unique elements of `ret_value` then mutex is not needed, especially before the loop.
Currently is serializes almost all the computation.